depends/hosts/linux.mk

```diff
@@ -33,6 +33,11 @@ x86_64_linux_AR=ar
 x86_64_linux_RANLIB=ranlib
 x86_64_linux_NM=nm
 x86_64_linux_STRIP=strip
+
+ifeq ($(NO_HARDEN),)
```

Comment:
I don't think this is the right place for these flags. If we want them in the Guix build, then we should add them to the Guix build, not hardcode them into depends.

Reply:
>  without controlling ... dependency versions

I might be misunderstanding you, but one of the main reasons depends exists is to control dependency versions? i.e as-opposed to using whatever ships with the system packages.

> so without controlling compiler

This change is a good example of trying to add that control; i.e hardcoding compiler flags with no check for if they are actually supported/or work as expected.

> Can you elaborate on how you think we should in general decide where configuration should live (default in configure/cmake vs depends vs guix)?

Sure! I'll follow up with something lengthier tomorrow.

Reply:
> Your concern here is that this may not work (or not work as expected) on the full range of environments where we expect depends-builds to work?

Yea. It's also somewhat displayed by the inconsistency of the change here; `C/XXFLAGS` is being put into depends, but the `LDFLAGS` are put into Guix. The LDFLAGS can't be hard-coded into depends, because that'll break depends builds where the libc/other base libs haven't been compiled with support for CET (for example, the current release of Alpine Linux on `x86_64`), or, any other builds where we use other libs compiled without such support, fuzzing, sanitizers etc.

Doing so would also add a requirement on the version of the linker, and possibly which ones
you can use. For any given flag, `ld`,`lld`, `mold` etc may/or may not implement it, and
will do so at various times (same holds for GCC vs Clang and compilation flags).

The `C/CXXFLAGS` addition might currently be ok (at least, nothing we test in CI is broken), and we know these flags are atleast "supported" by our minimum GCC and Clang, but now we're just playing whackamole with where things end up, and have got the inconsistency of spreading the flags for a certain feature across the two different systems, because after enough shuffling it now compiles and (hopefully) doesn't break anything.

Other concerns:

- Generally, this approach of hardcoding specific flags for specific targets (while simeltaneous just assuming they'll work) doesn't scale well: i.e for other `x86_64` targets where these compilation flags may be relevant; are we just copy-pasting the same blocks of flags around inside depends? That's messy.

- You've now got 3 different places to consider when thinking about hardening logic/flags and
their interactions (configure, depends, Guix), and, for anyone wanting to add new flags, it's even less
clear where they should live, or why.

> Can you elaborate on how you think we should in general decide where configuration should live
(default in configure/CMake vs depends vs Guix)?

Ideally, depends should remain generic/assumption free, and then users (including us), can use their distro of choice (which, for example, may already have it's own hardened toolchain), or bring their own compiler /
flags, and things will pretty much always just "work". We can then try flags generically/best effort in configure/CMake, and anyone can always continue to override what we are doing.

In that sense, you can think of Guix as essentially Cores own distro, where we have full
control to choose the compiler, glibc, binutils etc, configure them exactly as (and as hardened as)
we want, and then use that to compile depends and Core, with whichever additional flags we prefer for release
builds (that have additional post-build checks and ideally some coverage in CI).